### PR TITLE
CLI Metadata Plugin

### PR DESCRIPTION
**Pull request by manics**
Replacement for @joshmoore #3936

> For inspecting the availability of "metadata" (items shown in the right-hand panel of the GUIs) from the CLI, a `bin/omero metadata` plugin is being added along with tests to show how to add and query such data. The plugin also wraps populate_metadata.py and populate_roi.py mostly to simplify login.

Also:
- populate_metadata and populate_roi are now integrated into this CLI plugin (though calling the original scripts directly should still work)
- fixes for `populate_roi`
- commands for listing annotations
- commands for listing and deleting ROIs
- Initial implementation of https://trello.com/c/vcpdnS92/38-work-on-bulk-annotations-vs-map-annotations

Testing:
- `omero metadata --help`
- `omero metadata populate --file input.csv Screen:123`
- `omero metadata populate --file input.csv Plate:123`
- `omero metadata populate --context bulkmap` (uses bulk-annotation OMERO.table)
- `omero metadata populate --context bulkmap --file fileid`
- `omero metadata populateroi Plate:123`
- `omero metadata {summary,original,measures,bulkanns,mapanns,allannas,rois} ...`
- `omero metadata rois Object:Id`

- and for some sub-commands try the `--report` flag for additional output

@pwalczysko The fix for populate_roi should be in here thanks to @joshmoore

### CODE REVIEW

File: components/tools/OmeroPy/src/omero/plugins/metadata.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+
+import omero
+from omero.cli import BaseControl
+from omero.cli import CLI
+from omero.cli import ProxyStringType
+from omero.constants import namespaces
+from omero.gateway import BlitzGateway
+from omero.util import populate_metadata, populate_roi
+
+HELP = """metadata methods
```
Comment (ximenesuk) [inline components/tools/OmeroPy/src/omero/plugins/metadata.py]: The first help line looks a bit odd compared to most of our usual first lines. Would something like:

Metadata utitities

be better? (cf fs plugin)

File: components/tools/OmeroPy/src/omero/plugins/metadata.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        context_class = dict(self.POPULATE_CONTEXTS)[args.context]
+        ctx = context_class(client, args.obj, args.file)
+        ctx.parse()
+        if not args.dry_run:
+            ctx.write_to_omero()
+
+    def rois(self, args):
+        "Manage ROIs"
+        md = self._load(args)
+        if args.delete:
```
Comment (ximenesuk) [inline components/tools/OmeroPy/src/omero/plugins/metadata.py]: Would a dry-run option be useful for this subcommand option?

File: components/tools/OmeroPy/src/omero/plugins/metadata.py (hunk 1 above)
Comment (manics) [inline components/tools/OmeroPy/src/omero/plugins/metadata.py]: Fixed

File: components/tools/OmeroPy/src/omero/plugins/metadata.py (hunk 2 above)
Comment (manics) [inline components/tools/OmeroPy/src/omero/plugins/metadata.py]: Added for consistency, passed straight through to `omero delete`. In theory `omero rois Object:Id` should show the list of ROIs that would be deleted

**Comment by ximenesuk** [comment]
I have tried the workflow at https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/BulkAnnotations.html but using the cli without much luck.
```
ls30778:openmicroscopy colin$ omero metadata populate --file ~/Desktop/Bulk\ annotations/test_screen.csv Screen:1
Using session 54a20a55-d317-4db6-89ef-dc012c19b8d1 (user-0@localhost:4064). Idle timeout: 10 min. Current group: pg-0
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:Screen is missing plate: 365700055641
WARNING:omero.util.populate_metadata:PlateColumn is unimplemented
INFO:omero.util.populate_metadata:Created new table OriginalFile:767
INFO:omero.util.populate_metadata:Table initialized with 11 columns.
INFO:omero.util.populate_metadata:Added data column data.
ls30778:openmicroscopy colin$ omero metadata summary Screen:1
Using session 54a20a55-d317-4db6-89ef-dc012c19b8d1 (user-0@localhost:4064). Idle timeout: 10 min. Current group: pg-0
Screen:1
--------
Name: Screen1
Bulk annotations: 1
FileAnnotations: 1
```
There is an annotation on the Screen but not on individual wells. And with,
```
ls30778:openmicroscopy colin$ omero metadata populate --file ~/Desktop/Bulk\ annotations/test_plate.csv Plate:3
Using session 54a20a55-d317-4db6-89ef-dc012c19b8d1 (user-0@localhost:4064). Idle timeout: 10 min. Current group: pg-0
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
INFO:omero.util.populate_metadata:Missing image name column, skipping.
INFO:omero.util.populate_metadata:Missing plate name column, skipping.
Traceback (most recent call last):
  File "/Users/colin/Work/repos/openmicroscopy/dist/bin/omero", line 125, in <module>
    rv = omero.cli.argv()
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/cli.py", line 1437, in argv
    cli.invoke(args[1:])
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/cli.py", line 951, in invoke
    stop = self.onecmd(line, previous_args)
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/cli.py", line 1028, in onecmd
    self.execute(line, previous_args)
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/cli.py", line 1110, in execute
    args.func(args)
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/plugins/metadata.py", line 324, in populate
    ctx.write_to_omero()
  File "/Users/colin/Work/repos/openmicroscopy/dist/lib/python/omero/util/populate_metadata.py", line 572, in write_to_omero
    "Unable to create table: %s" % name)
omero.util.populate_metadata.MetadataError: Unable to create table: bulk_annotations
```
Is there anything else I need on my local branch? And @pwalczysko is that scenario and are the files used valid?

**Comment by pwalczysko** [comment]
@ximenesuk : The scenario was last tested during the Paris meeting this year. It worked up to the points which are documented in the tickets. One of the points in the tickets is that when you run the script against a Screen (as you did above in step 1) then the annotation on single wells is missing. When run against a plate this shoiuld work as of June this year. The scenario was not updated with respect to this PR (scenario is unaware of this PR). I do not think you should be missing anything on your branch. If your setup is wrong, in my experience you would not get accross to the output like 

```
NFO:omero.util.populate_metadata:Missing image name column, skipping.
```

**Comment by pwalczysko** [comment]
@ximenesuk : The "populate_roi.py does not work" ticket is https://trac.openmicroscopy.org/ome/ticket/12235

**Comment by ximenesuk** [comment]
@pwalczysko Note from the description that this PR fixes `populate_roi` though I have not yet tested that.

@manics do you have any sources of test data that should test the cli? And any guidance on `--context` ?

**Comment by pwalczysko** [comment]
@ximenesuk : This is the other ticket considering the populate_metadata.py - the two tickets pasted by me here fully describe all my knowledge about the status of this problematics (as indicated, my last test on this was done this year on the users' meeting in Paris) https://trac.openmicroscopy.org/ome/ticket/12211#comment:12

**Comment by manics** [comment]
`--context` is an exploratory option for some of the IDR work, so it may be removed/chagned before this ends up in develop (sorry for not describing it, I thought josh was around this week and would be reviewing it).

Essentially, 
`OMERO-TEST/bin/omero metadata populate --context bulkmap Object:Id` should convert a bulk-annotations table into a MapAnnotation on each image or well (depending on what was specified in the BulkAnnotations table). The MapAnnotations are in the bulk-annotations namespace. See the [Trello card](https://trello.com/c/vcpdnS92/38-work-on-bulk-annotations-vs-map-annotations) for the background.

The UI work for this has **not** been done- i.e. the table annotation is not removed, and annotations on wells aren't displayed, only those on images. #4038 is a temporary hack to force the display of Image and Well annotations.

`OMERO-TEST/bin/omero metadata populate --context deletemap Object:Id` should remove all MapAnnotations created as a result of the `bulkmap` context.

I'd expect the behaviour of `populate_metadata` to be unchanged in terms of what does/doesn't work. populate_roi should be fixed, at least for the test scenario.

**Comment by manics** [comment]
`omero metadata populate --report Plate:XX --file bulk_annotations/test_plate.csv` works for me, do you have any server error logs?

**Comment by manics** [comment]
And sources of data...just what's in the scenario. Assuming `populate_rois` works Well/Image A1 should have 173 ROIs, and you can use it to test the `bulkanns`/`measures` sub-commands with {`--parents`,`--report`} options
`omero metadata bulkanns Image:ID --parents --report`

**Comment by ximenesuk** [comment]
Okay, so after a restart of the server the bulk annotation of a Plate worked. However, the webclient on my local server is taking 15 seconds to respond to every request so I think the earlier problem is likely a local issue unless there is anything else in the metadat branch that might complicate things. I'll continue later after a laptop restart.

**Comment by manics** [comment]
You're welcome to test against dogfish

**Comment by ximenesuk** [comment]
I think I have gone as far as I can with this one without spending a bit more time catching up on the IDR side of things. I've tested the basic listing and populating using `--file` but not any of the `--context` stuff. No problems other than the flakiness of my server.

**Comment by manics** [comment]
I've got another PR building on top of this, further extending the plugin, and it's staying on the `metadata` branch for now. I think the important thing as far as testing is concerned is it doesn't break the existing populate_* functionality (or anything else).

**Comment by manics** [comment]
@joshmoore Over to you for further comments (note I'm already building on this in #4071 so if you prefer I could close this one and have the whole plugin reviewed at once)

**Comment by joshmoore** [comment]
Various issues while testing: (I'll update this comment)

 * `o metadata original Screen:2` fails
 * `o metadata summary Screen:1` fails if it doesn't exist

**Comment by manics** [comment]
Python 2.6 error fixed in #4062 53c0d5a
First two points in [above comment](https://github.com/openmicroscopy/openmicroscopy/pull/4037#issuecomment-137676030) fixed in #4062 08faf6f

**Comment by manics** [comment]
What do you prefer, close this PR and just review #4062 as one big PR? Or merge this (without fixes) since it doesn't break anything?

**Comment by manics** [comment]
Regarding the RFE `metadata populate --attach`, if you can attach the CSV this implies you should also be able to create the bulk-annotations from a previously uploaded OriginalFile which is fine.

And, if `populate_metadata.py` has significant work done on it, should `__main__` be removed so `omero metadata populate` is the only way to run it?

**Comment by manics** [comment]
@pwalczysko Regarding this PR, would you have any objections to me removing `populate_metadata/roi.py` in a future PR in 5.2 and only supporting `omero metadata populate ...`?

**Comment by pwalczysko** [comment]
@manics : assuming that ``omero metadata populate...`` does at least the same scope or even more as did ``populate_(metadata, roi).py``, I think this is fine. The only question would be whether we know about places which are totally dependent on the old scripts functioning ? @joshmoore @sbesson

**Comment by joshmoore** [comment]
@pwalczysko : I assume will finish that conversation when this comes back to develop, perhaps for 5.2?

Merging for SiB.

cc: @jburel @ximenesuk

**Pull request merged**